mlflow/sklearn/__init__.py

```diff
@@ -884,12 +944,59 @@ def patched_fit(original, self, *args, **kwargs):
                           for autologging (e.g., specify "fit" in order to indicate that
                           `sklearn.linear_model.LogisticRegression.fit()` is being patched)
         """
+        global _autolog_training_status
+        with _SklearnTrainingSession(clazz=self.__class__, allow_children=False) as t:
+            if t.should_log():
+                _autolog_training_status = _AutologTrainingStatus(
+                    last_mlflow_run_id=mlflow.active_run().info.run_id,
+                    model_id=id(self)
+                )
+                print(f'DBG: run patched_fit,'
+                      f'run_id={_autolog_training_status.last_mlflow_run_id},'
+                      f'model_id={_autolog_training_status.model_id}.')
+                try:
+                    _autolog_training_status.in_fit_call_scope = True
+                    result = fit_mlflow(original, self, *args, **kwargs)
+                finally:
+                    _autolog_training_status.in_fit_call_scope = False
+                return result
+            else:
+                return original(self, *args, **kwargs)
+
+    def patched_predict(original, self, *args, **kwargs):
+        global _autolog_training_status
         with _SklearnTrainingSession(clazz=self.__class__, allow_children=False) as t:
             if t.should_log():
-                return fit_mlflow(original, self, *args, **kwargs)
+                result = original(self, *args, **kwargs)
+                _autolog_training_status.register_predition_result(self, result)
+                return result
             else:
                 return original(self, *args, **kwargs)
 
+    def patched_metric_api(original, *args, **kwargs):
+        # TODO:
+        #  check _SklearnTrainingSession and skip logging if in nested runs.
+        global _autolog_training_status
+        metric = original(*args, **kwargs)
+
+        print('DGB: run patched_metric_api #1')
+        if not _autolog_training_status.in_fit_call_scope and \
+                not _autolog_training_status.in_scorer_call_scope:
```

Comment:
Here excludes patch under 2 cases
* metric API being called in the scope of `model.fit/fit_predict/fit_transform`, this case it is for computing the training score, so we need exclude patch under the case
* metric API being called in the scope of `scorer.__call__` (`scorer` is the object returned by `sklearn.metrics.make_scorer`), this case we have separate patching code to handle. See https://github.com/mlflow/mlflow/pull/4491/files#r656253501

Reply:
Can we use `_SklearnTrainingSession`?